app/gui/view/graph-editor/src/lib.rs

```diff
@@ -3652,3 +3661,171 @@ impl display::Object for GraphEditor {
+        // Adding a new node.
+        let (node_1_id, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port_shape().expect("No output port.");
+        port.events.mouse_down.emit(());
+        port.events.mouse_up.emit(());
+        assert_eq!(graph_editor.edges().len(), 1);
+        // Dropping edge.
+        let mouse = &app.display.default_scene.mouse;
+        let click_pos = Vector2(300.0, 300.0);
+        mouse.frp.position.emit(click_pos);
+        let click_on_background = |_: &GraphEditor| mouse.click_on_background();
+        let (_, node_2) = graph_editor.add_node_by(&click_on_background);
+        graph_editor.assert(Case { node_source: Some(node_1_id), should_edit: true });
+        let node_pos = node_2.position();
+        assert_eq!(node_pos.xy(), click_pos);
+    }
+
+    #[test]
+    fn test_connecting_two_nodes() {
+        let (_, ref graph_editor) = init();
+        let edges = graph_editor.edges();
+        assert!(graph_editor.nodes().is_empty());
+        assert!(edges.is_empty());
+        // Adding two nodes.
+        let (node_id_1, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        let (node_id_2, node_2) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port_shape().expect("No output port.");
+        port.events.mouse_down.emit(());
+        port.events.mouse_up.emit(());
+        let edge_id = graph_editor.on_edge_add.value();
+        let edge = edges.get_cloned_ref(&edge_id).expect("Edge was not added.");
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), None);
+        assert_eq!(edges.len(), 1);
+        // Connecting edge.
+        // We need to enable ports. Normally it is done by hovering the node.
+        node_2.model.input.frp.set_ports_active(true, None);
+        let port = node_2.model.input_port_shape().expect("No input port.");
+        port.hover.events.mouse_down.emit(());
+        port.hover.events.mouse_up.emit(());
+        assert_eq!(edge.source().map(|e| e.node_id), Some(node_id_1));
+        assert_eq!(edge.target().map(|e| e.node_id), Some(node_id_2));
+    }
+
+
+    // === Test utilities ===
+
+    /// An assertion case used when adding new nodes. See [`GraphEditor::assert`] below.
+    struct Case {
+        /// A source node of the added node.
+        node_source: Option<NodeId>,
+        /// Should we start the node editing immediately after adding it?
+        should_edit: bool,
+    }
+
+    impl GraphEditor {
+        fn add_node_by<F: Fn(&GraphEditor)>(&self, add_node: &F) -> (NodeId, Node) {
+            add_node(self);
+            let (node_id, ..) = self.node_added.value();
+            let node = self.model.nodes.all.get_cloned_ref(&node_id).expect("Node was not added.");
```

Comment:
[nit] do I understand correctly we can now shorten this line too?

```suggestion
            let node = self.nodes().get_cloned_ref(&node_id).expect("Node was not added.");
```